junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestableFactory.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine.junit5.descriptor;
+
+import static org.junit.gen5.commons.util.ReflectionUtils.loadClass;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.gen5.commons.util.Preconditions;
+import org.junit.gen5.commons.util.ReflectionUtils;
+
+public class JUnit5TestableFactory {
+
+	private static final String SEPARATORS = ":$#";
+
+	public JUnit5Testable fromUniqueId(String uniqueId, String engineId) {
+		Preconditions.notEmpty(uniqueId, "uniqueId must not be empty");
+		List<String> parts = split(uniqueId);
+		Preconditions.condition(parts.remove(0).equals(engineId), "uniqueId must start with engineId");
```

Comment:
Only the launcher can guarantee uniqueness since it sees all ids, an engine sees only its own. But again, let's stop working in this corner until we can discuss it with the rest.

Von meinem iPad gesendet

> Am 07.11.2015 um 00:58 schrieb Stefan Bechtold <notifications@github.com>:
> 
> In junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestableFactory.java:
> 
> > +import java.util.ArrayList;
> > +import java.util.Arrays;
> > +import java.util.List;
> > +import java.util.stream.Collectors;
> > +
> > +import org.junit.gen5.commons.util.Preconditions;
> > +import org.junit.gen5.commons.util.ReflectionUtils;
> > +
> > +public class JUnit5TestableFactory {
> > +
> > +	private static final String SEPARATORS = ":$#";
> > +
> > +	public JUnit5Testable fromUniqueId(String uniqueId, String engineId) {
> > +		Preconditions.notEmpty(uniqueId, "uniqueId must not be empty");
> > +		List<String> parts = split(uniqueId);
> > +		Preconditions.condition(parts.remove(0).equals(engineId), "uniqueId must start with engineId");
> I don't understand your question. Engines should not know about each others, at all.
> 
> Still the engine ID is part of the unique ID of a descriptor (at least for me it is a must-have). With this in place, we easily guarantee that descriptors created by different engines will never conflict each other. If we had unique ids without the engine id in it, we cannot guarantee uniqueness over several engines out-of-the-box. For me, the engine must be responsibility for uniqueness...
> 
> Please see my pull request.
> 
> —
> Reply to this email directly or view it on GitHub.
>